File: bulkSender.js
/**
 * Bulk Sender
 * Runs a pool of concurrent "workers" (Promises, not threads).
 * Each worker loops: pop a batch → send FCM → update stats → repeat.
 *
 * Concurrency math for 20M users in 60 seconds:
 *   20,000,000 / 500 tokens     = 40,000 batches
 *   40,000 / 60s                = ~667 batches/second needed
 *   500 concurrent × 150ms avg  = 3,333 batches/second capacity per thread
 *   4 threads × 3,333           = ~13,000 batches/second (20× headroom)
 */

import Redis from 'ioredis';
import { Config }        from './config.js';
import { initFirebase, sendMulticast } from './fcmClient.js';
import { StatsTracker }  from './stats.js';
import { popBatch }      from './tokenManager.js';

/**
 * @typedef {Object} NotificationPayload
 * @property {string} title
 * @property {string} body
 * @property {string} [image]
 * @property {Record<string,string>} [data]
 */

/**
 * Run one "worker" coroutine that continuously pops batches and sends them.
 * Stops when the queue is empty.
 *
 * @param {import('ioredis').Redis} redis
 * @param {NotificationPayload} notification
 * @param {StatsTracker} tracker
 */
async function worker(redis, notification, tracker) {
  while (true) {
    const tokens = await popBatch(redis, Config.BATCH_SIZE);

    // Empty queue → this worker is done
    if (tokens.length === 0) return;

    const result = await sendMulticast(tokens, notification);

    // Record stats
    await tracker.increment(result.successCount, result.failureCount);

    // Persist failed tokens for later retry
    if (result.failedTokens.length > 0) {
      const pipeline = redis.pipeline();
      for (const t of result.failedTokens) {
        pipeline.rpush(Config.REDIS_FAILED_QUEUE_KEY, t);
      }
      await pipeline.exec();
    }
  }
}

/**
 * Spin up `numWorkers` concurrent worker coroutines in this thread.
 * All workers share a single Redis connection and Firebase app.
 *
 * @param {NotificationPayload} notification
 * @param {number} totalTokens
 * @param {string} firebaseAppName - unique name per worker thread
 */
export async function runWorkers(notification, totalTokens, firebaseAppName) {
  // Each worker thread gets its own Firebase + Redis instance
  initFirebase(firebaseAppName);

  const redis = new Redis({
    host:            Config.REDIS_HOST,
    port:            Config.REDIS_PORT,
    db:              Config.REDIS_DB,
    maxRetriesPerRequest: 3,
    lazyConnect:     false,
  });

  const tracker = new StatsTracker(redis, totalTokens);

  // Launch CONCURRENT_WORKERS promises in parallel
  const workers = Array.from(
    { length: Config.CONCURRENT_WORKERS },
    () => worker(redis, notification, tracker)
  );

  await Promise.all(workers);
  await redis.quit();
}

File: bulk_sender.py
"""
Bulk Sender
Async worker pool that drains the Redis token queue and fires
FCM multicast requests at maximum throughput.

Math for 20M users in 60 seconds:
  - 20_000_000 / 500 (batch size)  = 40,000 batches
  - 40,000 / 60s                   = ~667 batches/second
  - With 500 concurrent workers,
    each request only needs to take < 750ms on average to hit the target.
"""

import asyncio
import logging
import time
from typing import Optional

import aioredis

from config import Config
from fcm_client import FCMClient, NotificationPayload
from stats import StatsTracker
from token_manager import pop_batch

logger = logging.getLogger(__name__)


async def worker(
    worker_id: int,
    redis: aioredis.Redis,
    client: FCMClient,
    notification: NotificationPayload,
    tracker: StatsTracker,
    semaphore: asyncio.Semaphore,
    stop_event: asyncio.Event,
):
    """
    A single async worker that:
    1. Pops a batch of tokens from Redis
    2. Sends multicast FCM request
    3. Updates stats
    4. Retries failed tokens once
    """
    while not stop_event.is_set():
        async with semaphore:
            tokens = await pop_batch(redis, Config.BATCH_SIZE)
            if not tokens:
                # Queue is empty, signal stop after brief pause
                await asyncio.sleep(0.05)
                stop_event.set()
                return

            # Send with retry
            result = None
            for attempt in range(Config.MAX_RETRIES):
                result = await client.send_multicast(tokens, notification)
                if result.failure_count == 0 or attempt == Config.MAX_RETRIES - 1:
                    break
                # Retry only failed tokens
                tokens = result.failed_tokens
                await asyncio.sleep(Config.RETRY_BACKOFF_BASE * (2 ** attempt))

            await tracker.increment(result.success_count, result.failure_count)

            # Store persistently-failed tokens for later analysis
            if result.failed_tokens:
                async with redis.pipeline(transaction=False) as pipe:
                    for t in result.failed_tokens:
                        pipe.rpush(Config.REDIS_FAILED_QUEUE_KEY, t)
                    await pipe.execute()

            logger.debug(
                "Worker %d: sent=%d failed=%d in %.0fms",
                worker_id,
                result.success_count,
                result.failure_count,
                result.duration_ms,
            )


async def run_workers(
    redis: aioredis.Redis,
    notification: NotificationPayload,
    tracker: StatsTracker,
    num_workers: int = Config.CONCURRENT_WORKERS,
):
    """
    Spin up *num_workers* async workers sharing a single FCMClient session.
    Uses a semaphore to cap concurrency and an event to signal completion.
    """
    client = FCMClient()
    await client.start()

    semaphore = asyncio.Semaphore(num_workers)
    stop_event = asyncio.Event()

    tasks = [
        asyncio.create_task(
            worker(i, redis, client, notification, tracker, semaphore, stop_event)
        )
        for i in range(num_workers)
    ]

    try:
        await asyncio.gather(*tasks)
    finally:
        await client.close()
